# Going Deeper into a regression model with a project

I hope you remember the Deep learning life cycle.

First step is loading the data.

## Loading the data

I'll be using the `Medical cost personal` dataset from the [kaggle](https://www.kaggle.com/datasets/mirichoi0218/insurance). 

This dataset contains personal health insurance costs for 1338 individuals. The columns are as follows:

- age: age of primary beneficiary
- sex: insurance contractor gender, female, male
- bmi: Body mass index, providing an understanding of body fatness
- children: Number of children covered by health insurance / Number of dependents
- smoker: Smoking, yes or no
- region: the beneficiary's geographic information, Southeast, Southwest, Northeast, Northwest
- charges: Individual medical costs billed by health insurance

A small and nice data set to clean and do some deep learning experiments.

You can download the dataset from kaggle or you can directly load from github without any downloading.

> Pandas can load raw data from links. Here's the [link](https://raw.githubusercontent.com/RishatTalukder/Deep_Learning_With_Tensorflow/refs/heads/main/2_regression/insurance.csv)

In [1]:
import pandas as pd

data = pd.read_csv(
    "https://raw.githubusercontent.com/RishatTalukder/Deep_Learning_With_Tensorflow/refs/heads/main/2_regression/insurance.csv"
)
data.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


See? nice right?

We can now use this dataset. 

So, time for the second step. 

##  Data Analysis

First let's check the information of the dataset.

In [2]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 94.5 KB


We have a total of 1338 rows and 7 columns and no missing values.

That makes our work simpler and I'm not thinking of going too deep, I'm going to keep it direct.

There's might be some categorical features in this dataset. Let's check that.

In [3]:
data.describe()

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


Only `age`, `bmi`, `children` and `charges` are numerical features. 

So, the rest can be categorical.  Let's see..

In [4]:
data.smoker.unique()

<ArrowStringArray>
['yes', 'no']
Length: 2, dtype: str

In [5]:
data.sex.unique()

<ArrowStringArray>
['female', 'male']
Length: 2, dtype: str

In [6]:
data.region.unique()

<ArrowStringArray>
['southwest', 'southeast', 'northwest', 'northeast']
Length: 4, dtype: str

Yeah, looks like `smoker`, `sex` and `region` features are categorical features.

There's another feature that can be categorical even if it is numerical.

In [7]:
data.children.unique()

array([0, 1, 3, 2, 5, 4])

Yeap, as you can see the children column can be categorical. But as this features represents the number of children, I'm going to keep it numerical.

Now, we have to transform the categorical feature to numetical feature. Because these feature are in string format.

So, now we can `one hot encode` the categorical features.

> You can leanr `One-hot encoding` from [here](https://youtu.be/FDLZNDyW3T4)

A little summery of `One-hot encoding`:

- One-hot encoding is a technique used to convert categorical variables into numerical variables.

- It creates a new column for each unique category in the categorical variable.

- Each cell in the new column is either 0 or 1, indicating whether the corresponding category is present in the original categorical variable.

- The value 1 indicates the presence of the category, while 0 indicates its absence.

- Let's say we have a categorical column like this:

| sex |
| --- |
| male |
| male |
| female |
| female |
| male |

- we takes all the unique values of the column and create a new column for each unique value.

| sex_male | sex_female |
| --- | --- |
| 1 | 0 |
| 1 | 0 |
| 0 | 1 |
| 0 | 1 |
| 1 | 0 |

- Each of these columns have a value of 1 or 0, indicating whether the corresponding category is present in the original categorical variable.

- Occasionally, we can drop one of these columns, because in this case if a row has 1 in the `sex_male` column, it means that the row has the `male` category, there is no chance that the row can have the `female` category. So, we can drop the `sex_female` column. This logic can be applied to any number of categories and we can drop a column and still the ammount of information will stay the same.


So, how do we  do this?

We can Either use a `pandas dataframe method` or a `sklearn preprocessing class` to do this.

I prefer to use both, which ever is convenient for me.

I show you both.

### One hot encoding using `get_dummies()`

In [8]:
pd.get_dummies(
    data,
    columns=["sex", "smoker", "region"],
    drop_first=True,
)

,age,bmi,children,charges,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,16884.92400,False,True,False,False,True
1,18,33.770,1,1725.55230,True,False,False,True,False
2,28,33.000,3,4449.46200,True,False,False,True,False
3,33,22.705,0,21984.47061,True,False,True,False,False
4,32,28.880,0,3866.85520,True,False,True,False,False
...,...,...,...,...,...,...,...,...,...
1333,50,30.970,3,10600.54830,True,False,True,False,False
1334,18,31.920,0,2205.98080,False,False,False,False,False
1335,18,36.850,0,1629.83350,False,False,False,True,False
1336,21,25.800,0,2007.94500,False,False,False,False,True


`pd.get_dummies` is a function that takes a pandas dataframe as input and returns a new dataframe with one-hot encoded columns for the specified categorical columns.

You can specify which columns to one-hot encode using the `columns` parameter or if you dont it'll one-hot encode all column with `object`, `string` or `category` dtype.

If you want to drop one of the one-hot encoded columns, you can use the `drop_first` parameter.

We can do the same using `sklearn` `OneHotEncoder` class.

In [9]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(
    sparse_output=False,
    drop="first",
)
encoder.set_output(transform="pandas")
encoder.fit_transform(data[["sex", "smoker", "region"]])

,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,0.0,1.0,0.0,0.0,1.0
1,1.0,0.0,0.0,1.0,0.0
2,1.0,0.0,0.0,1.0,0.0
3,1.0,0.0,1.0,0.0,0.0
4,1.0,0.0,1.0,0.0,0.0
...,...,...,...,...,...
1333,1.0,0.0,1.0,0.0,0.0
1334,0.0,0.0,0.0,0.0,0.0
1335,0.0,0.0,0.0,1.0,0.0
1336,0.0,0.0,0.0,0.0,1.0


Here I'm using `OneHotEncoder` class from `sklearn` to do the same thing.

Inside, I've set `sparse_output=False` because I want the output to be a pandas dataframe and not a sparse matrix.

And I've set `drop="first"` because I want to drop one of the one-hot encoded columns.

I've set `transform="pandas"` in using the `set_output` method because I want the output to be a pandas dataframe.

And after that I'm using `fit_transform` method to fit the encoder to the data and transform the data.

And we have the one-hot encoded dataframe.

I'll set up a pipeline for this process in this article later. But for now let's stick with `get_dummies` method.

In [10]:
data = pd.get_dummies(
    data,
    columns=["sex", "smoker", "region"],
    drop_first=True,
)

In [11]:
data.head()

,age,bmi,children,charges,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,16884.92400,False,True,False,False,True
1,18,33.770,1,1725.55230,True,False,False,True,False
2,28,33.000,3,4449.46200,True,False,False,True,False
3,33,22.705,0,21984.47061,True,False,True,False,False
4,32,28.880,0,3866.85520,True,False,True,False,False


Nice.

Now, we can just train a model. But first we have to split.

## Splitting the data

In this data set the target is to predict the `charges` column which represents the individual medical costs billed by health insurance.

In [12]:
X = data.drop(
    columns=["charges"],
)

y = data["charges"]

Now  we split.

In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
)

In [14]:
X_train.shape

(1070, 8)

And we have our train and test sets.

One thing left to do.

## Building the model and training it

I'll make a very small model for the first try.

In [15]:
import tensorflow as tf
from keras import (
    layers,
    losses,
    metrics,
    models,
    optimizers,
)

2026-09-02 12:31:47.606368: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [25]:


model = models.Sequential([
    layers.Input(shape=(X_train.shape[1],)),
    layers.Dense(10, activation="relu"),
    layers.Dense(5, activation="relu"),
    layers.Dense(1),
])

model.compile(
    optimizer=optimizers.Adam(),
    loss=losses.MeanAbsoluteError(),
    metrics=[metrics.RootMeanSquaredError()],
)

This model will have 3 layers, the first one will have 10 nodes, the second one will have 5 nodes and the last one will have 1 node.

And I'm using `Adam` optimizer and `MeanAbsoluteError` loss function and `RootMeanSquaredError` metric to see if the models are overfitting or underfitting.

In [26]:
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_9 (Dense)                 │ (None, 10)             │            90 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 5)              │            55 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 1)              │             6 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 151 (604.00 B)

 Trainable params: 151 (604.00 B)

 Non-trainable params: 0 (0.00 B)

Let's see how our model performs.

In [27]:
model.fit(
    X_train,
    y_train,
    epochs=100
)

Epoch 1/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - loss: 13330.4062 - root_mean_squared_error: 17944.4453
Epoch 2/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13317.6211 - root_mean_squared_error: 17934.3340 
Epoch 3/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13302.0215 - root_mean_squared_error: 17921.9141 
Epoch 4/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13283.4023 - root_mean_squared_error: 17907.0176 
Epoch 5/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13259.5645 - root_mean_squared_error: 17887.9316 
Epoch 6/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13228.2002 - root_mean_squared_error: 17863.1523 
Epoch 7/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13187.3027 - root_mean_squared_error: 17830.1621 
Epoch 8/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13133.8428 - root_mean_squared_error: 17788.7266 
Epoch 9/100
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 13065.2061 - root_mean_squared_error: 17734.3828 
Epoch 10/100
34/34 

Alright, no good. The loss is still pretty `High` and the RMSE shows that the model is predicting some pretty bad values.

But let's just evaluate it on the testing data first.

In [28]:
model.evaluate(X_test, y_test)

9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - loss: 7014.0020 - root_mean_squared_error: 12583.2295


[7014.001953125, 12583.2294921875]

Doesn't seem so bad right. But If we see that mean and  median of the target variable we might  know how bad this model is.

In [29]:
y.mean(), y.median()

(np.float64(13270.422265141257), 9382.033)

The mean absolute error is the avgerage of the absolute difference between the predicted and actual values. 

So, on average the model is off by 7014 dollers whereas the median is 13270 dollars. This is bad.

And if you take the median it's even worse.

So, we need to improve the model.

How do we do that.

We can do:

- Make the model more complex
- Train for longer.
- Try different optimizers.

Let's start with making the model more complex.

### Making the model more complex

This time let's go for 3 hidden layers and 64, 32, 16 neurons respectively.

In [ ]:
model2 = models.Sequential(
    [
        layers.Input(shape=(X_train.shape[1],)),
        layers.Dense(64, activation="relu"),
        layers.Dense(32, activation="relu"),
        layers.Dense(16, activation="relu"),
        layers.Dense(1),
    ]
)

model2.compile(
    optimizer=optimizers.Adam(),
    loss=losses.MeanAbsoluteError(),
    metrics=[metrics.RootMeanSquaredError()],
)

model2.fit(
    X_train,
    y_train,
    epochs=100
)